src/app_model/backends/qt/_qaction.py

```diff
@@ -118,52 +120,70 @@ class QMenuItemAction(QCommandRuleAction):
 
     Mostly the same as a `CommandRuleAction`, but aware of the `menu_item.when` clause
     to toggle visibility.
-    """
-
-    _cache: ClassVar[dict[tuple[int, int], QMenuItemAction]] = {}
 
-    def __new__(
-        cls: type[QMenuItemAction],
-        menu_item: MenuItem,
-        app: Application | str,
-        parent: QObject | None = None,
-        *,
-        cache: bool = True,
-    ) -> QMenuItemAction:
-        """Create and cache a QMenuItemAction for the given menu item."""
-        app = Application.get_or_create(app) if isinstance(app, str) else app
-        key = (id(app), hash(menu_item))
-        if cache and key in cls._cache:
-            return cls._cache[key]
+    Parameters
+    ----------
+    menu_item : MenuItem
+        `MenuItem` instance to create an action for.
+    app : Application | str
+        Application instance or name of application instance.
+    parent : QObject | None
+        Optional parent widget, by default None
+    """
 
-        self = super().__new__(cls)
-        if cache:
-            cls._cache[key] = self
-        return self  # type: ignore [no-any-return]
+    _cache: ClassVar[WeakValueDictionary[tuple[int, int], QMenuItemAction]] = (
+        WeakValueDictionary()
+    )
 
     def __init__(
         self,
         menu_item: MenuItem,
         app: Application | str,
         parent: QObject | None = None,
-        *,
-        cache: bool = True,  # used in __new__
     ):
-        initialized = False
-        with contextlib.suppress(RuntimeError):
-            initialized = getattr(self, "_initialized", False)
-
-        if not initialized:
-            super().__init__(menu_item.command, app, parent)
-            self._menu_item = menu_item
-            key = (id(self._app), hash(menu_item))
-            self.destroyed.connect(lambda: QMenuItemAction._cache.pop(key, None))
-            self._app.destroyed.connect(lambda: QMenuItemAction._cache.pop(key, None))
-            self._initialized = True
-
+        super().__init__(menu_item.command, app, parent)
+        self._menu_item = menu_item
         with contextlib.suppress(NameError):
             self.update_from_context(self._app.context)
 
+    @staticmethod
+    def _cache_key(app: Application, menu_item: MenuItem) -> tuple[int, int]:
+        return (id(app), hash(menu_item))
+
+    def _remove_from_cache(self) -> None:
+        cache_key = QMenuItemAction._cache_key(self._app, self._menu_item)
+        QMenuItemAction._cache.pop(cache_key, None)
```

Comment:
I'm not too worried about this coverage miss. it's obviously a result of not using a lambda (probably just never gets called in tests anymore).  All good to go otherwise?

Reply:
Hm. When running under debug, it looks that it is really uncalled. 
It looks like `obj.destroyer` is not called because qt do not call self signals on destroyed model. And `app.destroyed` is not called because Qt objects are deleted first. 
So it seems like WeakValueDictionary is enough for self-cleaning.

Reply:
> It looks like obj.destroyer is not called because qt do not call self signals on destroyed model. And app.destroyed is not called because Qt objects are deleted first.

right, I think it was only the usage of a lambda that was making it get called before this PR.  I assumed that's exactly what you were referring to when you said "as qt will not try to trigger the method of destroyed objects." :joy:

> So it seems like WeakValueDictionary is enough for self-cleaning?

I'm absolutely willing to give it a try